File: crons/cron3.js
var	_ = require('underscore'),
	async = require('async'),
	queue = require('payparrot_models/libs/queue'),
	Accounts = require('payparrot_models/objects/accounts.js'),
	Suscriptions = require('payparrot_models/objects/suscriptions.js'),
	Notifications = require('payparrot_models/objects/notifications.js'),
	request = require('request'),
	db = require('payparrot_models/libs/mongodb').connect();


// TODO: Manage errors
notify = function(notification_message, callback){
	var query_data = {
		suscription_id:"",
		account_id:"",
		parrot_id:"",
		type:"",
		external_id: ""
	};
	Suscriptions.findOne(
		{
			account_id: notification_message.Body.account_id,
			parrot_id: notification_message.Body.parrot_id
		}, {},function (err, suscription){
			Accounts.findOne({_id: notification_message.Body.account_id}, {account_id: 0}, function (err, account){

				var notification = new Notifications();

				switch(notification_message.Body.type){
					case "payment_success":
						console.log("payment_success");
						query_data.type = "payment_success";
						notification.type = query_data.notification_type;
						break;
					case "payment_failed":
						console.log("payment_failed");
						query_data.type = "payment_failed";
						notification.type = query_data.notification_type;						
						break;
					case "suscription_activated":
						console.log("suscription_activated");
						query_data.type = "suscription_activated";
						notification.type = query_data.notification_type;						
						break;
					case "suscription_deactivated":
						console.log("suscription_deactivated");
						query_data.type = "suscription_deactivated";
						notification.type = query_data.notification_type;						
						break;
					default:
						console.log("default");
				};

				if (account != null && typeof account.notification_url != "undefined") {
					// request.post?

					notification.account_id = account._id;
					notification.external_id = suscription.external_id;
					notification.parrot_id = suscription.parrot_id;
					notification.request_url = account.notification_url;

					query_data.suscription_id = suscription.id;
					query_data.account_id = notification.account_id;
					query_data.parrot_id = notification.parrot_id;
					query_data.external_id = notification.external_id;

					request.post(
						{
							url: account.notification_url,
							json: query_data
						}, 
						function (error, response, body) {
							notification.response_status = response.statusCode;
							notification.response_headers = response.headers;
							notification.response_body = response.body;
							notification.save(function(){
								queue.deleteMessage('notifications', notification_message.ReceiptHandle, function(err){
									callback();
								});
							});
						}
					);					
				} else {
					console.log("Notification couldn't be sent");
					callback();
				}
			});
		}
	);
};

var has_messages = true;

async.whilst(
    function () { return has_messages; },
    function(callback){
		queue.getMessage('notifications', function(message){
			if(message){
				console.log("message");
				async.forEach([message], notify, function(result){
					callback();
				});
			}else{
				console.log("no_message");
				has_messages = false;
				callback();
			}
		});
	},
    function (err) {
		if(err){
			console.log("An error ocurred");
			console.log(err);
		}
		db.connection.close();
    }
);

File: scripts/payparrot_scripts/crons/cron3.py
# -*- coding: utf-8 -*-
import sys
import json

from urllib import urlencode
from httplib2 import Http

from payparrot_dal import Subscriptions, Notifications, Accounts
from payparrot_dal.mongodb import connect
from payparrot_dal.queue import Queue
from payparrot_scripts.crons.utils import log 

VALID_NOTIFICATIONS = [
    "payment_success",
    "payment_failed",
    "payment_failed",
    "subscription_activated",
    "subscription_deactivated"
]

def main():
    connection = None
    # TODO: change i
    i = 0
    log('cron3', 'Starting')
    try:
        connection, db = connect()
        message = Queue.get_message('notifications')
        while message is not None and i<20:
            notification_message = json.loads(message.get_body())
            notify(db, message, notification_message)
            message = Queue.get_message('notifications')
            i+=1
    finally:
        if connection:
            connection.close()
        log('cron3', 'Finishing')

def notify(db, notification_raw, notification_message):
    notification = Notifications.findOne(db, notification_message.get('notification_id'))
    account = Accounts.findOne(db, notification.account_id)
    if notification:
        log('cron3', 'Notifying remote customer', notification_message.get('subscription_id'))
        if not notification_message.get('type') in VALID_NOTIFICATIONS:
            log('cron3', 'ERROR: Unknown notification', notification_message.get('subscription_id'))
            # TODO: Check what to do in this case
            return
        if notification.request_url:
            query_data = {
                'subscription_id': str(notification.subscription_id),
                'account_id': str(notification.account_id),
                'parrot_id': str(notification.parrot_id),
                'type': notification_message.get('type'),
                'external_id': str(notification.external_id),
                'notification_id': str(notification.id),
            }
            log('cron3', 'Notification URL: %s' % notification.request_url, notification_message.get('subscription_id'))
            utf8_query_data = dict([(key,val.encode('utf-8')) for key, val in query_data.items() if isinstance(val, basestring)])
            delete_message = False
            try:
                if account.notification_active:
                    http_client = Http()
                    headers, body = http_client.request(uri = notification.request_url, body = urlencode(utf8_query_data), method = 'POST')
                    if int(headers.status) >= 200 and int(headers.status) < 300:
                        notification.update({
                            'response_status': headers.status,
                            'response_headers': headers,
                            'response_body': body,
                            'status': 'sent'
                        })
                        log('cron3', 'Remote notification succeded', notification_message.get('subscription_id'))
                    else:
                        log('cron3', "Failed. Notification response not 2XX (received %s) from url %s" % (
                            headers.status,
                            notification.request_url
                        ), notification_message.get('subscription_id'))
                else:
                    delete_message = True
                    notification.update({
                            'status': 'off'
                    })
                if delete_message:
                    Queue.delete_message('notifications', notification_raw)

            except Exception, e:
                log('cron3', "Failed. Exception %specified" % e)
        else:
            log('cron3', 'ERROR: No remote url specified', notification_message.get('subscription_id'))
    else:
        log('cron3', "Failed. No notification found in db with id %s" % notification_message.get('notification_id'), notification_message.get('subscription_id'))


if __name__ == '__main__':
    main()        